// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";
option java_outer_classname = "HelloWorldProto";
option objc_class_prefix = "HLW";

package helloworld;

enum ErrorCode {
  SUCCESS = 0;
  FAILED = 1;
}

// The greeting service definition.
service Greeter {
  // Read data
  rpc ReadChunk (ReadChunkRequest) returns (ReadChunkReply) {}
  // Write data
  rpc WriteChunk (WriteChunkRequest) returns (WriteChunkReply) {}
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}

// Read Chunk Request
message ReadChunkRequest {
  int64 chunk_id = 1;
}

// Write Chunk Request
message WriteChunkRequest {
  int64 chunk_id = 1;
  string data = 2;
// TODO: hash of the data
}

// Read Chunk Reply
message ReadChunkReply {
  string data = 2;
// TODO: hash of the data
}

// Write Chunk Reply
message WriteChunkReply {
  ErrorCode error_code = 1;
}
